instrumentor/sources_webhooks.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/odigos-io/odigos/k8sutils/pkg/consts"
+	"github.com/odigos-io/odigos/k8sutils/pkg/workload"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type SourcesDefaulter struct {
+	client.Client
+}
+
+var _ webhook.CustomDefaulter = &SourcesDefaulter{}
+
+func (s *SourcesDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	source, ok := obj.(*v1alpha1.Source)
+	if !ok {
+		return fmt.Errorf("expected a Source but got a %T", obj)
+	}
+
+	if source.Labels == nil {
+		source.Labels = make(map[string]string)
+	}
+
+	if _, ok := source.Labels[consts.WorkloadNameLabel]; !ok {
+		source.Labels[consts.WorkloadNameLabel] = source.Spec.Workload.Name
+	}
+	if _, ok := source.Labels[consts.WorkloadNamespaceLabel]; !ok {
+		source.Labels[consts.WorkloadNamespaceLabel] = source.Spec.Workload.Namespace
+	}
+	if _, ok := source.Labels[consts.WorkloadKindLabel]; !ok {
+		source.Labels[consts.WorkloadKindLabel] = string(source.Spec.Workload.Kind)
+	}
+
+	if !v1alpha1.IsWorkloadExcludedSource(source) &&
+		source.DeletionTimestamp.IsZero() &&
+		!controllerutil.ContainsFinalizer(source, consts.DeleteInstrumentationConfigFinalizer) {
+		controllerutil.AddFinalizer(source, consts.DeleteInstrumentationConfigFinalizer)
+	}
+	if v1alpha1.IsWorkloadExcludedSource(source) &&
+		source.DeletionTimestamp.IsZero() &&
+		!controllerutil.ContainsFinalizer(source, consts.StartLangDetectionFinalizer) {
+		controllerutil.AddFinalizer(source, consts.StartLangDetectionFinalizer)
+	}
+
+	return nil
+}
+
+type SourcesValidator struct {
+	client.Client
+}
+
+var _ webhook.CustomValidator = &SourcesValidator{}
+
+func (s *SourcesValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+	source, ok := obj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", obj)
+	}
+
+	errs := s.validateSourceFields(ctx, source)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil, nil
+	}
+
+	return nil, apierrors.NewInvalid(
+		schema.GroupKind{Group: "odigos.io", Kind: "Source"},
+		source.Name, allErrs)
+}
+
+func (s *SourcesValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+	old, ok := oldObj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", old)
+	}
+	new, ok := newObj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", new)
+	}
```

Comment:
nit: if this error ends up hitting, the 2 lines has the same text, which can make it difficult for us later on to understand which of the 2 caused it.

will we benefit from adding `old` and `new` to the error text? or will we get stacktrace here?

Reply:
updated